### ISSUE TITLE
Texture atlases with offset don't work properly in UI

### ISSUE THREAD

**Issue by rparrett** [bevyengine/bevy#11219]
## Bevy version

0.12.1
recent main (0275508 is the last recent commit where texture atlases in UI aren't broken by a different bug)

## Relevant system information

```
AdapterInfo { name: "Apple M1 Max", vendor: 0, device: 0, device_type: IntegratedGpu, driver: "", driver_info: "", backend: Metal }
SystemInfo { os: "MacOS 13.6.2 ", kernel: "22.6.0", cpu: "", core_count: "10", memory: "64.0 GiB" }
```
`scale_factor` of `1.0` and `2.0` were tested.

## What you did

I first noticed this during Bevy Jam 4 as a slight offset in some of my UI images. Below you can see an example, with adjacent sprites showing at the right and bottom edges of the `E` and the right edge of the `F`. This is 103x50 texture atlas with 1px of offset and padding. This seemed to get worse at the sprites get further right/down in the sheet.

![bleed](https://github.com/bevyengine/bevy/assets/200550/8ee23429-072d-4611-a521-ea5f00490084)

Here's an example that shows the same exact `TextureAtlas` both as a sprite and in the UI.

```rust
use bevy::prelude::*;

fn main() {
    App::new()
        .add_plugins(DefaultPlugins.set(ImagePlugin::default_nearest()))
        .add_systems(Startup, setup)
        .add_systems(Update, increment_atlas_index)
        .run();
}

fn setup(
    mut commands: Commands,
    asset_server: Res<AssetServer>,
    mut texture_atlas_layouts: ResMut<Assets<TextureAtlasLayout>>,
) {
    commands.spawn(Camera2dBundle::default());

    let texture_handle = asset_server.load("sheettest.png");
    let texture_atlas = TextureAtlasLayout::from_grid(
        UVec2::new(16, 16),
        3,
        4,
        Some(UVec2::splat(16)),
        Some(UVec2::splat(40)),
    );
    let texture_atlas_handle = texture_atlas_layouts.add(texture_atlas);

    commands.spawn(AtlasImageBundle {
        style: Style {
            width: Val::Px(256.),
            height: Val::Px(256.),
            ..default()
        },
        image: texture_handle.clone().into(),
        texture_atlas: TextureAtlas::from(texture_atlas_handle.clone()),
        ..default()
    });

    commands.spawn(SpriteSheetBundle {
        texture: texture_handle,
        atlas: TextureAtlas::from(texture_atlas_handle),
        transform: Transform::from_scale(Vec3::splat(16.0)),
        ..default()
    });
}

fn increment_atlas_index(
    mut atlases: Query<&mut TextureAtlas>,
    keyboard: Res<ButtonInput<KeyCode>>,
) {
    let frames = 12;
    if keyboard.just_pressed(KeyCode::Space) {
        for mut atlas in &mut atlases {
            atlas.index = (atlas.index + 1) % frames;
        }
    }
}
```

And `sheettest.png`. It has exaggerated offset/padding with those areas highlighted for debugging purposes.

![sheettest](https://github.com/bevyengine/bevy/assets/200550/b9be3616-1e8e-46f7-845c-6d5ede20999e)

## What went wrong

Both images should look the same. Instead, we see this:

<img width="1280" alt="Screenshot 2024-01-04 at 2 05 26 PM" src="https://github.com/bevyengine/bevy/assets/200550/615c2611-a76d-408e-aceb-c3bb6fed229b">

The image in the middle is the sprite and looks correct. 

## Additional information

This works as expected with a texture atlas without offset/padding such as the one used in the `ui_texture_atlas` example.

**Comment by rparrett** [comment]
I believe the issues is either in

1. `TextureAtlas::from_grid` not calculating size as expected.
2. `extract_atlas_uinodes` using `TextureAtlas::size` instead of the size of the actual texture.

I'm not really sure what a "correct" fix would be.

`1` seems slightly problematic if `TextureAtlas::size` is supposed to be exactly the texture size, because the `offset` param is ambiguous about whether the texture has the "offset area" on its right and bottom sides as well.

`2` seems problematic, because after #10520, we can no longer access the actual texture size where it is needed.

So I think the right call is probably to fix `1` by adding `2. * offset` to size in `TextureAtlas::from_grid`. But I am very unfamiliar with UI rendering so there may be some better solution.

There is a workaround for end users, which is to set the `TextureAtlas`'s size to the correct value after it is created.

**Comment by rparrett** [comment]
After #11212, fixing `2` is on the table. But it would be good to understand what `TextureAtlas::size` is expected to be and if that's also something that needs fixing.

**Comment by outergod** [comment]
I've run into the same problem. I checked the `Rect`s stored in `TextureAtlas::textures` by `from_grid`, the calculated areas look correct based on padding and offset - which makes sense, given that `SpriteSheetBundle` renders the correct sprite for each tested index.

I did notice however that in case of `AtlasImageBundle`, the original texture looks scaled to match the calculated height in `TextureAtlas::size`, in other words, unwanted scaling is applied before obtaining the sprites from the texture, which explains padding and offsets not matching. If I set the `tile_size` to match the texture height in `TextureAtlas::from_grid`, the rendering result from `AtlasImageBundle` looks identical.
